crates/matrix-sdk-crypto/src/types/events/room_key_withheld.rs

```diff
@@ -0,0 +1,718 @@
+        }
+    }
+
+    /// Get the withheld code of this event.
+    pub fn withheld_code(&self) -> WithheldCode {
+        match self {
+            RoomKeyWithheldContent::MegolmV1AesSha2(c) => c.withheld_code(),
+            #[cfg(feature = "experimental-algorithms")]
+            RoomKeyWithheldContent::MegolmV2AesSha2(c) => c.withheld_code(),
+            RoomKeyWithheldContent::Unknown(c) => c.code.to_owned(),
+        }
+    }
+
+    /// Get the algorithm of the room key withheld.
+    pub fn algorithm(&self) -> EventEncryptionAlgorithm {
+        match &self {
+            RoomKeyWithheldContent::MegolmV1AesSha2(_) => EventEncryptionAlgorithm::MegolmV1AesSha2,
+            #[cfg(feature = "experimental-algorithms")]
+            RoomKeyWithheldContent::MegolmV2AesSha2(_) => EventEncryptionAlgorithm::MegolmV2AesSha2,
+            RoomKeyWithheldContent::Unknown(c) => c.algorithm.to_owned(),
+        }
+    }
+}
+
+impl EventType for RoomKeyWithheldContent {
+    const EVENT_TYPE: &'static str = "m.room_key.withheld";
+}
+
+/// A machine-readable code for why the megolm key was not sent.
+#[derive(
+    Eq,
+    Hash,
+    PartialEq,
+    Clone,
+    AsStrAsRefStr,
+    AsRefStr,
+    FromString,
+    DebugAsRefStr,
+    SerializeAsRefStr,
+    DeserializeFromCowStr,
+)]
+#[non_exhaustive]
+pub enum WithheldCode {
+    /// the user/device was blacklisted.
+    #[ruma_enum(rename = "m.blacklisted")]
+    Blacklisted,
+
+    /// the user/devices is unverified.
+    #[ruma_enum(rename = "m.unverified")]
+    Unverified,
+
+    /// The user/device is not allowed have the key. For example, this would
+    /// usually be sent in response to a key request if the user was not in
+    /// the room when the message was sent.
+    #[ruma_enum(rename = "m.unauthorised")]
+    Unauthorised,
+
+    /// Sent in reply to a key request if the device that the key is requested
+    /// from does not have the requested key.
+    #[ruma_enum(rename = "m.unavailable")]
+    Unavailable,
+
+    /// An olm session could not be established.
+    /// This may happen, for example, if the sender was unable to obtain a
+    /// one-time key from the recipient.
+    #[ruma_enum(rename = "m.no_olm")]
+    NoOlm,
+
+    #[doc(hidden)]
+    _Custom(PrivOwnedStr),
+}
+
+impl std::fmt::Display for WithheldCode {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        let string = match self {
+            WithheldCode::Blacklisted => "The sender has blocked you.",
+            WithheldCode::Unverified => "The sender has disabled encrypting to unverified devices.",
+            WithheldCode::Unauthorised => "You are not authorised to read the message.",
+            WithheldCode::Unavailable => "The requested key was not found.",
+            WithheldCode::NoOlm => "Unable to establish a secure channel.",
+            _ => self.as_str(),
+        };
+
+        write!(f, "{string}")
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug)]
```

Comment:
```suggestion
#[derive(Debug, Deserialize, Serialize)]
```

std derives first, for consistency